### ISSUE TITLE
Include Exit Numbers and expose them in RouteStep

### ISSUE THREAD

**Issue by roman-yepishev** [Project-OSRM/osrm-backend#3400]
Based on #228 routing engine supposedly can provide information on exit number, however currently the engine tells "take the ramp on the right" or "keep right at the fork".

I was unable to find any references for motorway_junction in the code, so I assume the functionality to relay the exit numbers from OSM is not present.

Is it something OSRM plans to include?

**Comment by daniel-j-h** [comment]
We handle exit / destination tags for the `destination=` key on ways:

https://wiki.openstreetmap.org/wiki/Key:destination

We don't handle `exit_to=` tags on nodes:

https://wiki.openstreetmap.org/wiki/Key:exit_to

mostly because the node tag is a historic mistake used primarily for rendering purpose and hard to work around on our side. Closing as duplicate of https://github.com/Project-OSRM/osrm-backend/issues/3317.

**Comment by roman-yepishev** [comment]
Hm, how to get OSRM to provide the exit _number_?AFAIR ref= on the highway=motorway_junction is the exit number and destination tag is not for setting the exit number.On Dec 4, 2016 15:52, "Daniel J. H." <notifications@github.com> wrote:We handle exit / destination tags for the destination= key on ways:
https://wiki.openstreetmap.org/wiki/Key:destination
We don't handle exit_to= tags on nodes:
https://wiki.openstreetmap.org/wiki/Key:exit_to
mostly because the node tag is a historic mistake used primarily for rendering purpose and hard to work around on our side. Closing as duplicate of #3317.

**Comment by danpat** [comment]
@roman-yepishev you can't right now, we don't extract the necessary tags, nor do we support an "exit number" attribute in the API response.  We only support destination tags so far.

@daniel-j-h I think we should consider re-opening this ticket as a separate enhancement to #3317.  Exit numbers and destinations aren't the same thing, we might actually want both in the response for highway ramps.  At least in the US, exits are numbered and navigation like "Take exit 33 onto the I95 towards Spokane" (Take exit `exit_to` onto the `ref/name` towards `destination:ref`).

http://wiki.openstreetmap.org/wiki/Proposed_features/junction_details

Near me, they seem to be tagged on junction nodes with `highway=motor_junction` and `ref=NNN`.  Example:

https://www.mapillary.com/app/?lat=47.637103&lng=-117.473553&z=17&pKey=FeCWcVDAWrVk4HfK_F4vcw&focus=photo

Note the "Exit 277" sign.

http://www.openstreetmap.org/node/50889156#map=17/47.63717/-117.47164&layers=ND

In this case, the expected instruction would be:

    Use exit 277 to get on Highway 2 West towards Spokane Airport

**Comment by daniel-j-h** [comment]
@danpat What you propose is taking the ref tag

http://wiki.openstreetmap.org/wiki/Key:ref

and in addition to what we're already doing (exposing it as `ref` in the response) you want a flag indicating if it's an exit, a simple road etc. - correct? The use case for this then would be announcing it as "take exit 277" instead of e.g. "take the ramp towards Spokane (277)".

Here's the current response:

![ref](https://cloud.githubusercontent.com/assets/527241/20879002/f299fff4-bad1-11e6-95e9-856ae54da006.png)


http://map.project-osrm.org/?z=17&center=47.638585%2C-117.468030&loc=47.639239%2C-117.468610&loc=47.637916%2C-117.474071&hl=en&alt=0

Due to the `destination:ref` tag on the way

http://www.openstreetmap.org/way/6146402

I'm fine with re-opening this if you want to.

**Comment by danpat** [comment]
@daniel-j-h yes, that's pretty much it.  We should have an indication in the response that the maneuver is a numbered highway-exit, and can be announced as such.

**Comment by danpat** [comment]
If we start with the API side, my basic proposal would be:

```json
  "maneuver": {
    "type": "off-ramp",
    "modifier": "right",
  },
  "exit-number": "277B"
  "ref": "US2-W",
  "destinations": [...]
```

where `exit-number` is an optional field that will only appear for `off-ramp` maneuvers.

**Comment by TheMarex** [comment]
Hrm I still don't fully grasp the difference between:

- `destination`
- `destination:ref`
- What you propose as `exit-number`

Looking at the linked OSM wiki page the exit number should be tagged as `junction:ref` on the link way and as `ref` on the `highway=motorway_junction` node.

So we don't actually need better `destination` coverage for exit numbers but porting all the node based `ref` values to `junction:ref`?

**Comment by daniel-j-h** [comment]
@danpat please clarify on Patrick's question.

cc @pratikyadav ticket for context.

**Comment by danpat** [comment]
@TheMarex At least in the US, the offramp from a highway has a few bits of information available to the driver:

  - the name of the road this ramp becomes - this is usually the `ref` tag on the ramp way
  - the general name of where this ramp will take you - this is usually the `destination` tag
  - the numbered "exit", i.e. "Exit 277" - this is usually the `ref` on the **node** at the highway exit point, not on the way itself.  For a similar thing, see https://github.com/Project-OSRM/osrm-backend/issues/3317 for `exit_to`

This offramp has an example:
http://www.openstreetmap.org/way/6146402

and the node on the highway where it exits has `ref=277`, which would be "Exit 277".  
http://www.openstreetmap.org/node/50889156

In the US at least, the exit numbers on the federal interstate roads are usually based on how many miles from the state border you are, but in general, they're just monotonically increasing - this makes it easy for drivers to think "Hey, I need exit 277, and I just went past 274 - I have (3 exits | 3 miles) until I need to turn".

This `exit number` property is unique to motorways - I don't ever recall seeing one on smaller roads, where `ref` and `destination` tagging still exists.

**Comment by emiltin** [comment]
in denmark, exits on motorways are also numbered sequentially

**Comment by TheMarex** [comment]
@danpat thank you for that explanation. 👍 

It seems from what you describe we should translate this node-based property to [`junction:ref`](https://wiki.openstreetmap.org/wiki/Proposed_features/junction_details) based on @daniel-j-h work for `exit_to` vs `destination`.

I would suggest using `exit` or `exit_ref` as field name since I'm not sure this will always be a number-like thing in every country of usage.

**Comment by danpat** [comment]
Yes, it's definitely not always numeric.  Even locally for me, values like `Exit 29A` and `Exit 29B` are common.  This needs to be stored as a string.

**Comment by daniel-j-h** [comment]
Update: we need to expose an exit number field in the [`RouteStep`](http://project-osrm.org/docs/v5.7.0/api/#routestep-object) for countries such as the US where exits are numbered and having them in the response is crucial for great highway user experience.

Think: "take the exit 29B towards YourCity".

We scoped out to support the [`junction:ref`](https://wiki.openstreetmap.org/wiki/Proposed_features/junction_details) tag - mostly because it is tagged on ways and this makes it possible for us to attach it to our internal data structures. Eventually we could re-write `ref` tags _on the motorway junction node_ into `junction:ref` tags on the way similar to https://github.com/mapbox/rewrite-exit-destination-signage.

**Comment by daniel-j-h** [comment]
Here's what needs to be done (ordered by how the osrm pipeline works). In general you can follow the `name`, `ref`, `pronunciation` and `destination` property we already have in place (e.g. see https://github.com/Project-OSRM/osrm-backend/pull/2857).
- [ ] Add a new field to our ExtractionWay
- [ ] Expose the new field to Lua in our ScriptingEnvironment
- [ ] In the profiles parse `junction:ref` in the `way_function` similar to how we parse `name` and `ref` and add the value to the `result` field you just exposed
- [ ] in extractor_callbacks.hpp adapt the string deduplication map
- [ ] in extractor_callbacks..cpp deduplicate and add your field to external_memory datastructure
- [ ] add a ref attribute to the RouteStep type representing a step in the response
- [ ] add function for returning the ref based on name id to all datafacades
- [ ] in assemble_steps.hpp call datafacade function and add the result to the RouteStep object
- [ ] in json_factory.cpp serialize the RouteStep's new attribute to the json::object
- [ ] adapt cucumber support code: route.js, shared_steps.js,
- [ ] add cucumber tests testing the new behavior
- [ ] add api docs for the new ref field in RouteStep to docs/http.md
- [ ] add a changelog entry for both profiles and api change

**Comment by TheMarex** [comment]
This was implemented in #4215.
